Feat gui rule editor

Hugo-Barroca opened the pull request:
Code for the rule editor pop-up. Some sections will have to be changed in the future, once all necessary classes and data structures have been created. The JUnit tests are not yet implemented, only the base structure is.

Hugo-Barroca committed:
Added a basic GUI for the Edit Rule popup.

Hugo-Barroca committed:
Reorganized the code, added functionality to some of the JButtons.

Hugo-Barroca committed:
Code reorganizations, JavaDocs created.

Hugo-Barroca committed:
General Editing

Added IFs to the metrics'list and a delete rule button, as well as a
base structure for future JUnit tests.

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,277 @@
+	 * Creates and returns the JPanel responsible for holding the JButtons which
+	 * allow all metrics to be cleared, or the rule to be saved.
+	 * 
+	 * @return
```
missing the return

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,277 @@
+		
+		clearButton.addActionListener(new ActionListener() {
+			public void actionPerformed(ActionEvent e) {
+				System.out.println("Clear button pressed!");
```
does this print appear anywhere? what's the purpose?

Hugo-Barroca replied:
Print has been deleted.

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,277 @@
+				if (nameText.getText().isEmpty()) {
+					JOptionPane.showMessageDialog(null, "Please insert a rule name!");
+				} else {
+					System.out.println("Delete button pressed!");
```
same applies here

Hugo-Barroca replied:
Print has been deleted.

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,277 @@
+					JOptionPane.showMessageDialog(null, "Please insert a rule name!");
+				} else {
+					System.out.println("Save button pressed!");
+					System.out.println("Rule name: " + nameText.getText());
```
and here :p

Hugo-Barroca replied:
Print has been deleted.

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,277 @@
+						metric = condition.getSelectedItem() + " IF " + value.getSelectedItem() + " "
+								+ comparison.getSelectedItem() + " " + threshold.getText();
+					}
+					System.out.println("Add metric button pressed!");
```
if these were only for debugging, please remove all prints

Hugo-Barroca replied:
They were indeed mostly for debugging. I kept them because I thought they might be useful later on, when adding functionality to this GUI, and since they're console prints, they wouldn't really be viewed by a regular user anyway. I'll delete for now though, if need be they're easy to add back in regardless.

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,277 @@
+		value.addItem("MET");
+		
+		JComboBox<String> comparison = new JComboBox<>();
+		comparison.addItem(">");
```
can we make an enum for the operators and one for the metrics and use those instead of hard coding this?

vandabarata replied:
Resolving this since it'll be done in a later task

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,277 @@
+		addMetricButton.addActionListener(new ActionListener() {
+			public void actionPerformed(ActionEvent e) {
+				if (value.getItemCount() == 0) {
+					JOptionPane.showMessageDialog(null, "Please check if all metric fields were entered correctly!");
```
seems like a weird thing to say, metrics aren't entered by the user, they're predetermined by us. I think this may be unnecessary?

Hugo-Barroca replied:
The metrics themselves are predetermined by us, but the threshold is user input, which is what this _if_ **should**  be guarding against. The _if_ as it currently stands is wrong, however. It served to check if the selection box had something selected in earlier stages in the development process. Since this is no longer a possibility, the if condition is serving no purpose. I'll remove it when I get the chance, and implement a proper validation for the user input.

vandabarata replied:
please update this when possible

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,277 @@
+								+ threshold.getText();
+						
+					} else {
+						metric = condition.getSelectedItem() + " IF " + value.getSelectedItem() + " "
```
gonna say, can we take the if out of the other rules? like, usually it'd be 'if blabla and whatever or whatever else'

vandabarata commented on src/main/java/gui/EditRulePopupTest.java:
```diff
@@ -0,0 +1,13 @@
+class EditRulePopupTest {
+	@Test
+	void test() {
+		fail("Not yet implemented");
```
I was thinking, for your test, just assert the size of the window being created, since you have that predefined

Hugo-Barroca replied:
Alright, I'll look into it.

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,277 @@
+	/**
+	 * Creates and returns the JPanel responsible for holding the rule's name.
+	 * 
+	 * @return
```
missing return

vandabarata replied:
You marked it as resolved but there's still nothing here or the other returns

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,277 @@
+	 * Creates and returns the JPanel responsible for holding both the current list
+	 * of rule metrics, and the line which allows the addition of more metrics.
+	 * 
+	 * @return
```
missing return

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,277 @@
+	 * hard-coded values of the JBoxes are to be replaced with the correct, final
+	 * values once the ENUMs are created.
+	 * 
+	 * @return
```
missing return

Hugo-Barroca commented:
![](https://github.trello.services/images/mini-trello-icon.png) [(5) [Task] Popup para adição/ edição de regras](https://trello.com/c/TNhoZTe2/13-5-task-popup-para-adi%C3%A7%C3%A3o-edi%C3%A7%C3%A3o-de-regras)

Hugo-Barroca committed:
Removed debug printlines

As per requested in pull request #7, I've deleted all unnecessary printlines.

Hugo-Barroca committed:
Took the Extra IFs out of the UI

Hugo-Barroca committed:
Fixed the previous commit's introduced bug.

Hugo-Barroca committed:
Added unit tests.

Hugo-Barroca commented:
All the requested changes have been (in theory) either implemented or addressed. Someone please review when possible, thanks!

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,283 @@
+		return addNewMetricPanel;
+	}
+
+//TODO
```
What is there to do here? you didn't write anything

Hugo-Barroca replied:
The TODO's accompanying test was written in the javadocs right bellow it. I will put the TODO in said javadocs por clarity.

vandabarata commented on src/main/java/gui/EditRulePopupMain.java:
```diff
@@ -0,0 +1,8 @@
+package main.java.gui;
+
+//This class was used for testing purposes and should be deleted once a proper way to call EditRulePopup is created.
```
Add a //TODO for easier tracking please

vandabarata commented on src/test/java/gui/TestEditRulePopup.java:
```diff
@@ -0,0 +1,17 @@
+import main.java.gui.EditRulePopup;
+
+class TestEditRulePopup {
+	@Test
```
please explain what's being done in the test, namely that you're validating the frame size to ensure the popup is being shown

vandabarata commented on src/test/java/gui/TestEditRulePopup.java:
```diff
@@ -0,0 +1,17 @@
+
+class TestEditRulePopup {
+	@Test
+	void test() {
```
name this appropriately please. @lrnsa-iscteiul may be able to help you with this

Hugo-Barroca replied:
Do we have any defined naming standart? I used the same name format as the main GUI's current testing class, that being test{ClassName}.

Hugo-Barroca replied:
I did not notice the test name, I was refering to the class name above. I've now changed the method name to be more descriptive of what the test is doing.

Hugo-Barroca committed:
Minor changes to comments and javadocs.

Hugo-Barroca committed:
Changed the test() method to testWindowSizeAtStart().

Hugo-Barroca committed:
Added the missing returns to the javadocs.

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	/**
+	 * Returns the popup's main SWING frame.
```
redundant, no point in saying exactly the same 2 lines apart in the same method - I'd remove this line

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	/**
+	 * Creates and returns the JPanel responsible for holding the rule's name.
```
same as above

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,285 @@
+	/**
+	 * 
+	 * Creates and returns the JPanel responsible for holding both the current list
+	 * of rule metrics, and the line which allows the addition of more metrics.
```
and again

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,285 @@
+
+	/**
+	 * Creates and returns the JPanel responsible for holding the JButtons which
+	 * allow all metrics to be cleared, or the rule to be saved.
```
oh come on!

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,285 @@
+	 * This method both returns and creates the JPanel holding the line which allows
+	 * users to add new metrics into the metrics list. TODO: The currently
+	 * hard-coded values of the JBoxes are to be replaced with the correct, final
+	 * values once the ENUMs are created.
```
Keep the TODO on its own line, otherwise it gets lost.
Also same as previous comments - no use in repeating the exact same stuff

Hugo-Barroca committed:
Cleaned up JavaDocs, introduced a basic form of user input validation
for the threshold.

franfaccin commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,274 @@
+	public EditRulePopup() {
+		frame = new JFrame("Personalized Rules");
+		frame.setLayout(new BorderLayout());
+		frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
```
This popup should not have this. It should not terminate the application on close. It should just close the window.
```suggestion
```

vandabarata replied:
Here's an example of how to change the close window listener to something else (taken from mine and fran's PCD project last year's):

frame.addWindowListener(new java.awt.event.WindowAdapter() {
			@Override
			public void windowClosing(WindowEvent e) {
				user.unsubscribeFromDiretorio();
				e.getWindow().dispose();
				System.exit(0);
			}
		});

vandabarata replied:
I imagine we want to confirm save on exit and when closing this popup, go back to the main frame

franfaccin replied:
I think we never close the previous window when opening this one. So it shouldn't be a problem.

franfaccin replied:
About the event listener thing, do we need to have a "on close" action? It will just close the pop up, so I don't think we need to attach anything to this event.

franfaccin commented on src/main/java/gui/EditRulePopupMain.java:
```diff
@@ -0,0 +1,8 @@
+package main.java.gui;
+
+//TODO: This class was used for testing purposes and should be deleted once a proper way to call EditRulePopup is created.
+public class EditRulePopupMain {
```
Please remove this class and trigger the `EditRulePopup` from the `MainFrame`

franfaccin commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,274 @@
+		JTextField threshold = new JTextField("10");
+
+		JButton addMetricButton = new JButton("Add");
+		addMetricButton.addActionListener(new ActionListener() {
```
You can use a lambda here, it's easier to read.

franfaccin commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,274 @@
+				} else {
+					metric = condition.getSelectedItem() + " " + value.getSelectedItem() + " "
+							+ comparison.getSelectedItem() + " " + threshold.getText();
+				}
```
Instead of repeating ` + comparison.getSelectedItem() + " " + threshold.getText();` in both cases, you can have
```java
String baseMetric =  comparison.getSelectedItem() + " " + threshold.getText();
if (ruleMetrics.isEmpty()) {
	metric = "IF " + value.getSelectedItem() + " " baseMetric;
} else {
	metric = condition.getSelectedItem() + " " baseMetric;
}
```
or if you want to be supper short:
```java
String baseMetric =  comparison.getSelectedItem() + " " + threshold.getText();
String metric = ruleMetrics.isEmpty()
    ? "IF " + value.getSelectedItem() + " " baseMetric
    : condition.getSelectedItem() + " " baseMetric;
```

franfaccin commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,274 @@
+		addNewMetricPanel.add(addMetricButton);
+		addNewMetricPanel.setPreferredSize(new Dimension(650, 25));
+		addNewMetricPanel.setMaximumSize(new Dimension(650, 25));
+		return addNewMetricPanel;
```
I don't know which part of the code does this, and this is just a nice to have. But it would be nice if the lines of the added rules would be centred in the middle of the panel, instead of being equally spaced.

<img width="777" alt="image" src="https://user-images.githubusercontent.com/14364481/69367189-b2edb180-0c8f-11ea-95ab-6ebb0b030c7f.png">
<img width="778" alt="image" src="https://user-images.githubusercontent.com/14364481/69367280-e3355000-0c8f-11ea-9078-cc44a765a2de.png">

vandabarata replied:
@ftfni-iscteiul open this as an increment for improvement - don't shoulder @Hugo-Barroca with this in this task x)

franfaccin replied:
ok

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,274 @@
+				if (nameText.getText().isEmpty()) {
+					JOptionPane.showMessageDialog(null, "Please insert a rule name!");
+				} else {
+					// Add functionality here.
```
```suggestion
					// TODO: Add functionality here.
```

vandabarata commented on src/main/java/gui/EditRulePopup.java:
```diff
@@ -0,0 +1,274 @@
+				if (nameText.getText().isEmpty()) {
+					JOptionPane.showMessageDialog(null, "Please insert a rule name!");
+				} else {
+					// Add functionality here.
```
```suggestion
					// TODO: Add functionality here.
```

vandabarata reviewed (changes requested):
Please address mine and @ftfni-iscteiul's comments when possible :)

Hugo-Barroca committed:
Deleted default close operation.

Co-Authored-By: Franciele Faccin <francielefaccin@gmail.com>

Hugo-Barroca committed:
Added TODO

Co-Authored-By: Vanda Barata <vsfba1@iscte-iul.pt>

Hugo-Barroca committed:
Added TODO

Co-Authored-By: Vanda Barata <vsfba1@iscte-iul.pt>

Hugo-Barroca committed:
Clarified setConditionVisibility's JavaDocs.

Hugo-Barroca committed:
Added an enumerator for metrics.

Hugo-Barroca committed:
Corrected the createAddMetricPanel

Corrected the createAddMetricPanel to work with the new metric
enumerator.

Hugo-Barroca committed:
Changed EditRulePopup (See description bellow for details).

Removed default threshold value for metric, added a Basic and Advanced
mode.

Hugo-Barroca committed:
Fixed a user input validation bug.

The previous validation would give a warning if the user input was
incorrect, but it would add the input in nonetheless. This has been
fixed.

Hugo-Barroca committed:
Made the changes from Basic mode reflect on the Advanced mode.

Hugo-Barroca committed:
Added Condition enums.

Hugo-Barroca committed:
Made the Basic mode button disabled by default.

Hugo-Barroca committed:
Updated some JavaDocs, added a ScrollPane to the advanced editing view.

Hugo-Barroca committed:
Merge branch 'master' into feat-gui-rule-editor.

Hugo-Barroca committed:
Deleted the EditRulePopupMain testing class, integrated with MainFrame.

franfaccin commented on src/main/java/controller/Condition.java:
```diff
@@ -0,0 +1,5 @@
+package main.java.controller;
```
Enum is not a controller, so should not be here. Either move to .model package or create an enums package.

Hugo-Barroca committed:
Moved the metric enum to the model package.

Hugo-Barroca committed:
Fixed a bug where the Condition JComboBox would reset.

Hugo-Barroca committed:
Shortened createMetricAddPanel() as request.

Hugo-Barroca committed:
Added output for the user on pressing the delete/save rule buttons.

vandabarata commented on src/main/java/model/Metric.java:
```diff
@@ -1,8 +1,8 @@
 /**
  * Metric enumerator, for convinience. The current existing metrics are LOC
- * (Lines of Code), CYCLO (McCabe�s Cyclomatic Number), ATFD (Access To Foreign
+ * (Lines of Code), CYCLO (McCabe�s Cyclomatic Number), ATFD (Access To Foreign
```
encoding looks off here, take a look when you can

Hugo-Barroca committed:
Fixed encoding issue in the Metric enum class.

vandabarata reviewed (changes requested):
Please add ENUM for operators and add lambda to action listener

Hugo-Barroca committed:
Added an enum for the Operators.

Hugo-Barroca committed:
Changed the Operators ENUM to Comparator, implemented it into the popup.

Hugo-Barroca committed:
Implemented lambda functions for improved readibility.

Hugo-Barroca committed:
Added JavaDocs, fixed typo.

Hugo-Barroca committed:
Merge branch 'master' into feat-gui-rule-editor

vandabarata commented on src/main/java/model/Comparator.java:
```diff
@@ -0,0 +1,22 @@
+ *
+ */
+public enum Comparator {
+	GT(">"), LT("<"), EQ("=="), DF("!=");
```
não vamos ter GE e LE (greater or equal, lesser or equal)?

vandabarata replied:
also da-lhe outro nome q nao Comparator - Operator would be better

Hugo-Barroca replied:
Eu tinha colocado Operator originalmente, mas como eles não fazem nenhuma operação tirando a comparação fiquei na dúvida... Vou mudar para Operator though, também acho que fica melhor.

vandabarata commented on src/test/java/gui/TestEditRulePopup.java:
```diff
@@ -0,0 +1,21 @@
+ * This test is validating the window size, to ensure the popup is being
+ * displayed properly.
+ */
+class TestEditRulePopup {
```
estava a pensar.. (podes meter só TODO e passar à frente, não te mates mais neste PR) - tendo em conta as funcionalidades novas, não há mais nada que possa ser coberto por mais unit tests?

Hugo-Barroca replied:
Not sure... Por um lado, diria que o advanced mode tem bem mais possibilidades de implementar unit tests do que o resto do código, já que uma vez que é texto corrido os resultados são mais imprevisiveis... Ao mesmo tempo, não faço nenhum tipo de verificação quanto ao input aqui, essa verificação vai ser feita mais do lado do código em que a fran está a trabalhar, então... I'll have to look into it. Apesar disso, em termos de UI tive uns bugs estranhos em relação às JComboBoxes duplicarem as opções selecionaveis. Esses bugs entretanto já foi corrigidos, mas verificar se o número de elementos selecionaveis é o correto em cada uma das combo boxes pode ser uma opção razoável para aumentar a cobertura do código.

vandabarata replied:
sounds great :D achas que dá para fazer isso rápido? :)

franfaccin replied:
Eu voto actually em o Hugo só fazer o essencial pra terminar esse PR. Ele já está a trabalhar desde sempre nisso e estamos precisando mto dessa ui pros próximos tickets. O "só mais isso aqui, só mais ali" já juntou mais 1 semana de trabalho pra ele. Also, o meu ticket é o de fazer o store das regras, então os unit tests disso acho que fazem parte do meu ticket. Lembrem que a ui não guarda regra de negócio. O que vai acontecer é que o meu controlador não vai deixar salvar um regra não válida.

vandabarata replied:
sim mas tal como ele disse, a lista tem de apresentar o número certo de rules, e isso é do UI e deve ser testado. Mais vale fazer aqui ou por um TODO para fazer aqui, do que deixar por fazer e passar

djsar-iscteiul replied:
Quando a parte lógica estiver toda implementada e só faltar os Tests Unit,  eu acredito que o melhor é fazer merge para podemos seguir com o trabalho. A parte do Hugo tem impacto e provoca dependencias com as próximas tarefas. Os testes fazemos quando as tarefas relacionadas estiverem implementadas. Eu concordo com a @ftfni-iscteiul que esta seja a melhor estratégia.

Hugo-Barroca committed:
Added two extra Comparators.

Hugo-Barroca committed:
Merge branch 'feat-gui-rule-editor' of https://github.com/vsfba1-iscteiul/ES1-2019-EIPL-81.git into feat-gui-rule-editor

Hugo-Barroca committed:
Changed the Operator enum and the MainFrame (see description).

Changed the Comparator enum's name to Operator and added two elementos
to it. Fixed an error on the Mainframe which was introduced by a small
edit on github.

Hugo-Barroca committed:
Added a small TODO in the TestEditRulePopup, created a few getters.

Hugo-Barroca committed:
Fixed small refactoring bug.

vandabarata reviewed (approved):
\o/
